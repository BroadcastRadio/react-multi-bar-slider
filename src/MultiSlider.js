import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';

export const transition = 'all 450ms cubic-bezier(.23, 1, .32, 1) 0ms';

export const ProgressSlider = glamorous.div({
  height: 14,
  boxSizing: 'border-box',
  transition
}, ({ readOnly, width, height, backgroundColor }) => ({
  width,
  height,
  backgroundColor,
  cursor: readOnly ? 'auto' : 'pointer'
}));

const Dot = glamorous.span({
  position: 'absolute',
  display: 'block',
  zIndex: 5,
  borderRadius: '50%'
}, ({ right, hasIcon, width, height, color, noTransition }) => ({
  top: hasIcon ? 0 : '50%',
  right: `${right || 0}%`,
  transform: hasIcon ? 'translateX(-50%)' : 'translate(50%, -50%)',
  width: hasIcon ? 0 : width,
  height: hasIcon ? 0 : height,
  backgroundColor: hasIcon ? 'transparent' : color,
  transition: noTransition ? 'none' : transition
}));

const DotIcon = glamorous.img({
  position: 'absolute',
  transform: 'translateX(-50%)',
  userSelect: 'none'
}, ({ width, height }) => ({
  width,
  height
}));

export const Progress = glamorous.div({
  position: 'absolute',
  top: 0,
  right: 0
}, ({ color, progress, height, equal, equalColor, zIndex, noTransition }) => ({
  width: `${progress || 0}%`,
  height,
  backgroundColor: equal && equalColor ? equalColor : color,
  zIndex,
  transition: noTransition ? 'none' : transition
}));

export const SlidableZone = glamorous.div({
  position: 'absolute',
  width: '100%'
}, ({ size, sliderCount }) => ({
  top: -size,
  bottom: -size,
  zIndex: sliderCount
}));

const limitProgress = progress => Math.max(Math.min(progress, 100), 0);

export default class MultiSlider extends Component {
  static propTypes = {
    sliders: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string.isRequired,
        progress: PropTypes.number.isRequired,
        dot: PropTypes.oneOfType([
          PropTypes.shape({
            color: PropTypes.string,
            icon: PropTypes.string,
            style: PropTypes.object
          }),
          PropTypes.bool
        ])
      }).isRequired
    ).isRequired,
    activeSlider: PropTypes.number,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    slidableZoneSize: PropTypes.number,
    backgroundColor: PropTypes.string,
    equalColor: PropTypes.string,
    sliderStyle: PropTypes.object,
    onSlide: PropTypes.func.isRequired,
    readOnly: PropTypes.bool
  };

  static defaultProps = {
    activeSlider: 0,
    width: '100%',
    height: 14,
    slidableZoneSize: 7,
    backgroundColor: '#EEEEEE',
    sliderStyle: {},
    readOnly: false
  };

  state = { mouseDown: false };

  handleSliderClick = (e) => {
    const { sliders, activeSlider, onSlide, readOnly } = this.props;

    if (readOnly) return;

    const sliderEl = this.sliderElement(e);
    const newProgress = this.progressFromMousePos(e, sliderEl);

    const oldProgress = sliders[activeSlider].progress;
    if (newProgress === oldProgress) return;

    onSlide(newProgress);
  };

  handleSliderMouseDown = () => this.setState({ mouseDown: true });
  handleSliderMouseUp = () => this.setState({ mouseDown: false });

  handleSliderMouseMove = (e) => {
    if (!this.state.mouseDown) return;
    this.handleSliderClick(e);
  };

  sliderElement = (e) => {
    const sliderEl = e.target;

    // Make sure we select the slider element
    if (sliderEl.classList[1] === 'icon') {
      return sliderEl.parentNode.parentNode;
    }
    if (sliderEl.classList[1] === 'dot' || sliderEl.classList[1] === 'progress') {
      return sliderEl.parentNode;
    }
    return sliderEl;
  };

  progressFromMousePos = (e, sliderEl) => {
    const boundingRect = sliderEl.getBoundingClientRect();
    const mouseFromLeft = e.pageX - boundingRect.left;
    const progressFromLeft = mouseFromLeft / boundingRect.width;
    // Return progress in percents
    return limitProgress(Math.round((1 - progressFromLeft) * 100));
  };

  sortSliders = () => {
    const sliders = [...this.props.sliders];
    return sliders.sort((a, b) => b.progress - a.progress);
  };

  slidersEqual = () => {
    const sliders = this.props.sliders;
    return sliders.reduce((acc, slider, i) => {
      if (!acc) return acc;
      if (!sliders[i - 1]) return true;
      if (slider.progress === sliders[i - 1].progress) return true;
      return false;
    }, true);
  };

  render = () => {
    const {
      sliders,
      width,
      height,
      slidableZoneSize,
      backgroundColor,
      equalColor,
      sliderStyle,
      readOnly
    } = this.props;
    const slidersEqual = this.slidersEqual();
    const sortedSliders = this.sortSliders();

    return (
      <ProgressSlider
        width={width}
        height={height}
        backgroundColor={backgroundColor}
        css={sliderStyle}
        readOnly={readOnly}
        onClick={this.handleSliderClick}
        onMouseDown={this.handleSliderMouseDown}
        onMouseUp={this.handleSliderMouseUp}
        onMouseMove={this.handleSliderMouseMove}
      >
        {sliders.map((slider, i) =>
          <Fragment key={i}>
            <Progress
              className="progress"
              color={slider.color}
              progress={slider.progress}
              height={height}
              equal={slidersEqual}
              equalColor={equalColor}
              zIndex={sortedSliders.indexOf(slider)}
              noTransition={this.state.mouseDown}
            />
            {slider.dot && (
              <Dot
                className="dot"
                right={slider.progress}
                hasIcon={!!slider.dot.icon}
                width={slider.dot.width || 28}
                height={slider.dot.height || 28}
                color={slider.dot.color || slider.color}
                css={slider.dot.style}
                noTransition={this.state.mouseDown}
              >
                {slider.dot.icon && (
                  <DotIcon
                    className="icon"
                    src={slider.dot.icon}
                    width={slider.dot.width || 50}
                    height={slider.dot.height || 50}
                    draggable={false}
                    css={slider.dot.iconStyle}
                  />
                )}
              </Dot>
            )}
          </Fragment>
        )}
        <SlidableZone size={slidableZoneSize} sliderCount={sliders.length} />
      </ProgressSlider>
    );
  };
}
