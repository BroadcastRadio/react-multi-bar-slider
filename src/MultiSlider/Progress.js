import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import sliderPropType from './utils/sliderPropType';
import glamorous from 'glamorous';
import transition from './utils/transition';
import getHalf from './utils/getHalf';
import Dot from './Dot';

const StyledProgress = glamorous.div(
  {
    position: 'absolute',
    top: 0,
    right: 0
  },
  ({
    color,
    progress,
    height,
    equal,
    equalColor,
    roundedCorners,
    zIndex,
    noTransition
  }) => ({
    width: `${progress || 0}%`,
    height,
    backgroundColor: equal && equalColor ? equalColor : color,
    borderRadius: roundedCorners ? getHalf(height) : 0,
    zIndex,
    transition: noTransition ? 'none' : transition
  })
);

const Progress = ({
  slider,
  height,
  slidersEqual,
  equalColor,
  roundedCorners,
  mouseDown,
  zIndex
}) => (
  <Fragment>
    <StyledProgress
      className="progress"
      color={slider.color}
      progress={slider.progress}
      height={height}
      equal={slidersEqual}
      equalColor={equalColor}
      roundedCorners={roundedCorners}
      zIndex={zIndex}
      noTransition={mouseDown}
    />
    {slider.dot && (
      <Dot
        dot={slider.dot}
        progress={slider.progress}
        sliderColor={slider.color}
        mouseDown={mouseDown}
      />
    )}
  </Fragment>
);

Progress.propTypes = {
  slider: sliderPropType,
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  slidersEqual: PropTypes.bool.isRequired,
  equalColor: PropTypes.string.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  mouseDown: PropTypes.bool.isRequired,
  zIndex: PropTypes.number.isRequired
};

export default Progress;
