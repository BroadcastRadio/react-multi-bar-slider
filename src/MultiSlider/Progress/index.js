import React from 'react';
import PropTypes from 'prop-types';
import sliderPropType from '../utils/sliderPropType';
import Dot from './Dot';
import processStyle from '../utils/processStyle';
import StyledProgress from './StyledProgress';

const Progress = ({
  slider,
  height,
  slidersEqual,
  equalColor,
  roundedCorners,
  mouseDown,
  zIndex
}) => (
  <StyledProgress
    className="progress"
    color={slider.color}
    progress={slider.progress}
    height={height}
    equal={slidersEqual}
    equalColor={equalColor}
    roundedCorners={roundedCorners}
    noTransition={mouseDown}
    zIndex={zIndex}
    css={processStyle(slider.style, {
      slider,
      height,
      slidersEqual,
      equalColor,
      roundedCorners,
      mouseDown,
      zIndex
    })}
  >
    {slider.dot && (
      <Dot
        dot={slider.dot}
        sliderColor={slider.color}
        mouseDown={mouseDown}
      />
    )}
  </StyledProgress>
);

Progress.propTypes = {
  slider: sliderPropType,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  slidersEqual: PropTypes.bool.isRequired,
  equalColor: PropTypes.string,
  roundedCorners: PropTypes.bool.isRequired,
  mouseDown: PropTypes.bool.isRequired,
  zIndex: PropTypes.number.isRequired
};

export default Progress;
