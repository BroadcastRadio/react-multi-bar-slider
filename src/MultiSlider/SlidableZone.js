import PropTypes from 'prop-types';
import styled from '@emotion/styled';

const SlidableZone = styled('div')({
  position: 'absolute',
  width: '100%'
}, ({ size, zIndex }) => ({
  top: -size,
  bottom: -size,
  zIndex
}));

SlidableZone.propTypes = {
  size: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  zIndex: PropTypes.number.isRequired
};

export default SlidableZone;
