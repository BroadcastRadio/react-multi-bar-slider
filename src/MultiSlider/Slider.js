import React from 'react';
import PropTypes from 'prop-types';
import processStyle from '../utils/processStyle';
import StyledSlider from './StyledSlider';

const Slider = ({
  width,
  height,
  backgroundColor,
  style,
  onMouseMoveActivate,
  onMouseMoveDeactivate,
  onMouseMove,
  roundedCorners,
  readOnly,
  children,
  ...props
}) => (
  <StyledSlider
    width={width}
    height={height}
    backgroundColor={backgroundColor}
    roundedCorners={roundedCorners}
    readOnly={readOnly}
    css={processStyle(style, {
      width,
      height,
      backgroundColor,
      roundedCorners,
      readOnly
    })}
    onMouseDown={onMouseMoveActivate}
    onMouseUp={onMouseMoveDeactivate}
    onMouseLeave={onMouseMoveDeactivate}
    onMouseMove={onMouseMove}
    {...props}
  >
    {children}
  </StyledSlider>
);

Slider.displayName = 'Slider';

Slider.propTypes = {
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  backgroundColor: PropTypes.string.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  onMouseMoveActivate: PropTypes.func.isRequired,
  onMouseMoveDeactivate: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired
};

export default Slider;
