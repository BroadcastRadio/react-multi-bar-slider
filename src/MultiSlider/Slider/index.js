import React from 'react';
import PropTypes from 'prop-types';
import processStyle from '../utils/processStyle';
import StyledSlider from './StyledSlider';

const Slider = ({
  width,
  height,
  backgroundColor,
  style,
  onSlide,
  onMouseMoveActivate,
  onMouseMoveDeactivate,
  onMouseMove,
  roundedCorners,
  readOnly,
  children
}) => (
  <StyledSlider
    width={width}
    height={height}
    backgroundColor={backgroundColor}
    roundedCorners={roundedCorners}
    readOnly={readOnly}
    css={processStyle(style, {
      width,
      height,
      backgroundColor,
      roundedCorners,
      readOnly
    })}
    onClick={onSlide}
    onMouseDown={onMouseMoveActivate}
    onMouseUp={onMouseMoveDeactivate}
    onMouseLeave={onMouseMoveDeactivate}
    onMouseMove={onMouseMove}
  >
    {children}
  </StyledSlider>
);

Slider.propTypes = {
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  backgroundColor: PropTypes.string.isRequired,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  onSlide: PropTypes.func.isRequired,
  onMouseMoveActivate: PropTypes.func.isRequired,
  onMouseMoveDeactivate: PropTypes.func.isRequired,
  onMouseMove: PropTypes.func.isRequired,
  roundedCorners: PropTypes.bool.isRequired,
  readOnly: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired
};

export default Slider;
