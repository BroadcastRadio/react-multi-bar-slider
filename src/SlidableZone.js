import PropTypes from 'prop-types';
import glamorous from 'glamorous';

const SlidableZone = glamorous.div({
  position: 'absolute',
  width: '100%'
}, ({ size, zIndex }) => ({
  top: -size,
  bottom: -size,
  zIndex
}));

SlidableZone.propTypes = {
  size: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  zIndex: PropTypes.number.isRequired
};

export default SlidableZone;
