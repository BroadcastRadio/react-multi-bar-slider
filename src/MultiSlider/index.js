import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import getProgressFromMousePosition from './utils/getProgressFromMousePosition';
import slidersEqual from './utils/slidersEqual';
import sortSliders from './utils/sortSliders';
import sliderPropType from './utils/sliderPropType';
import Slider from './Slider';
import Progress from './Progress';
import SlidableZone from './SlidableZone';

export default class MultiSlider extends PureComponent {
  static propTypes = {
    sliders: PropTypes.arrayOf(sliderPropType).isRequired,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    slidableZoneSize: PropTypes.number,
    backgroundColor: PropTypes.string,
    equalColor: PropTypes.string,
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    onSlide: PropTypes.func.isRequired,
    roundedCorners: PropTypes.bool,
    reverse: PropTypes.bool,
    readOnly: PropTypes.bool
  };

  static defaultProps = {
    width: '100%',
    height: 14,
    slidableZoneSize: 7,
    backgroundColor: '#EEEEEE',
    roundedCorners: false,
    style: {},
    reversed: false,
    readOnly: false
  };

  state = { mouseDown: false };

  handleSlide = (e) => {
    const { onSlide, reversed, readOnly } = this.props;
    if (readOnly) return;

    const newProgress = getProgressFromMousePosition(e, reversed);
    onSlide(newProgress);
  };

  handleMouseMoveActivate = () => this.setState({ mouseDown: true });
  handleMouseMoveDeactivate = () => this.setState({ mouseDown: false });

  handleMouseMove = (e) => {
    if (!this.state.mouseDown) return;
    this.handleSlide(e);
  };

  render = () => {
    const {
      sliders,
      width,
      height,
      slidableZoneSize,
      backgroundColor,
      equalColor,
      style,
      roundedCorners,
      reversed,
      readOnly
    } = this.props;
    const allSlidersEqual = slidersEqual(sliders);
    const sortedSliders = sortSliders(sliders);

    return (
      <Slider
        width={width}
        height={height}
        backgroundColor={backgroundColor}
        style={style}
        onSlide={this.handleSlide}
        onMouseMoveActivate={this.handleMouseMoveActivate}
        onMouseMoveDeactivate={this.handleMouseMoveDeactivate}
        onMouseMove={this.handleMouseMove}
        roundedCorners={roundedCorners}
        readOnly={readOnly}
      >
        {sliders.map((slider, i) =>
          <Progress
            slider={slider}
            height={height}
            slidersEqual={allSlidersEqual}
            equalColor={equalColor}
            roundedCorners={roundedCorners}
            reversed={reversed}
            mouseDown={this.state.mouseDown}
            zIndex={sortedSliders.indexOf(slider)}
            key={i}
          />
        )}
        <SlidableZone size={slidableZoneSize} zIndex={sliders.length} />
      </Slider>
    );
  };
}
