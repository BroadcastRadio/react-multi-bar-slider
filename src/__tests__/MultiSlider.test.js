import React from 'react';
import Enzyme, { shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import MultiSlider from '../MultiSlider';
import Slider from '../MultiSlider/Slider';
import SlidableZone from '../MultiSlider/SlidableZone';

Enzyme.configure({ adapter: new Adapter() });

beforeAll(() => {
  global.multiSliderProps = {
    sliders: [
      {
        color: '#00BDAF',
        progress: 17,
        dot: {
          icon: 'foo.png',
          iconStyle: { bottom: -7 }
        }
      },
      {
        color: '#AB47BC',
        progress: 45,
        style: ({ slidersEqual }) => ({
          borderTopLeftRadius: slidersEqual ? 7 : 0
        }),
        dot: {
          icon: 'bar.png',
          iconStyle: { top: 7 }
        }
      }
    ],
    slidableZoneSize: 50,
    onSlide: () => {}
  };
});

afterAll(() => {
  delete global.multiSliderProps;
});

const Child = () => <div />;

/* global multiSliderProps */
describe('MultiSlider.js', () => {
  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <MultiSlider {...multiSliderProps}>
          <Child progress={4} />
        </MultiSlider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const multiSlider = render(
      <MultiSlider {...multiSliderProps}>
        <Child progress={4} />
      </MultiSlider>
    );
    expect(multiSlider).toBeDefined();
  });

  it('renders a Slider', () => {
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps}>
        <Child progress={4} />
      </MultiSlider>
    );
    expect(multiSlider.find(Slider).length).toBe(1);
  });

  it('passes on undocumented props to the Slider', () => {
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps} foo="bar">
        <Child progress={4} />
      </MultiSlider>
    );
    expect(multiSlider.find(Slider).prop('foo')).toBe('bar');
  });

  it('passes the correct props to its Progress children', () => {
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps}>
        <Child progress={4} />
      </MultiSlider>
    );
    expect(multiSlider.find('Child').props()).toEqual({
      height: 14,
      progressEqual: true,
      equalColor: undefined,
      roundedCorners: false,
      reversed: false,
      mouseDown: false,
      zIndex: 0,
      progress: 4
    });
  });

  it('renders a SlidableZone', () => {
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps}>
        <Child progress={4} />
      </MultiSlider>
    );
    expect(multiSlider.find(SlidableZone).length).toBe(1);
  });

  it('handles slider clicks correctly', () => {
    const onSlide = jest.fn();
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps} onSlide={onSlide}>
        <Child progress={4} />
      </MultiSlider>
    );
    multiSlider.find(Slider).prop('onSlide')({
      target: {
        classList: [],
        getBoundingClientRect: () => ({
          left: 154,
          width: 876
        })
      },
      pageX: 933.64
    });
    expect(onSlide.mock.calls[0][0]).toBe(89);
  });

  it('handles mouse moves correctly', () => {
    const onSlide = jest.fn();
    const multiSlider = shallow(
      <MultiSlider {...multiSliderProps} onSlide={onSlide}>
        <Child progress={4} />
      </MultiSlider>
    );
    const fakeEvent = {
      target: {
        classList: [],
        getBoundingClientRect: () => ({
          left: 154,
          width: 876
        })
      },
      pageX: 933.64
    };
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls.length).toBe(0);

    multiSlider.find(Slider).prop('onMouseMoveActivate')();
    multiSlider.update();
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls[0][0]).toBe(89);

    multiSlider.find(Slider).prop('onMouseMoveDeactivate')();
    multiSlider.update();
    multiSlider.find(Slider).prop('onMouseMove')(fakeEvent);
    expect(onSlide.mock.calls.length).toBe(1);
  });
});
