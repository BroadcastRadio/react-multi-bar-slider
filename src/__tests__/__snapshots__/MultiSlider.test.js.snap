// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSlider.js matches the snapshot 1`] = `
<div
  className="css-pbel19"
  height={14}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  readOnly={false}
  width="100%"
>
  <div />
  <div
    className="css-1srlg56"
    size={50}
  />
</div>
`;
